[WIP]Added Phalcon\Mvc\Model\Binder

Jurigag opened the pull request:
PR with model binding changes in dispatcher, moving code to separated class, as well add more abilities, caching of reflection, and make binding working for micro handlers, controllers and lazy controllers
- [x] Make sure that the PR passes in Travis CI to make the process more efficient.
- [x] Only use tabs for indentation.
- [x] if needed, rebase to the proper branch before submitting your pull request.
    If it doesn't merge cleanly with master you may be asked to rebase your changes.
- [x] Don't put submodule updates in your pull request unless they are to landed commits.
- [x] Add tests relevant to the fixed bug or new feature. See our testing guide for
    more information.
- [x] Phalcon 2 is written in Zephir, please do not submit commits that modify C generated
    files directly or those whose functionality/fixes are implemented in the C
    programming language
- [x] Remove any change to ext/kernel / *.zep.c / *.zep.h files before submitting the PR

sergeyklay commented on phalcon/mvc/controller/bindmodelinterface.zep:
```diff
@@ -23,14 +23,17 @@ namespace Phalcon\Mvc\Controller;
-	abstract static function getModelName() -> string;
-}
\ No newline at end of file
+	abstract static function getModelName() -> var;
```
```
-> string | array
```

?

sergeyklay commented on phalcon/mvc/micro.zep:
```diff
@@ -690,16 +695,39 @@ class Micro extends Injectable implements \ArrayAccess
+
+					if realHandler instanceof Controller {
+						let methodName = handler[1];
+						let bindCacheKey = "binding_" . get_class(realHandler) . "_" . methodName;
```
```
"binding_" => "_PHMB_"
```

sergeyklay commented on phalcon/mvc/micro.zep:
```diff
@@ -1040,4 +1068,40 @@ class Micro extends Injectable implements \ArrayAccess
+	/**
+	 * Sets model binder
+	 *
+	 * @param \Phalcon\Mvc\Model\BinderInterface modelBinder
```
not needed

sergeyklay commented on phalcon/mvc/micro/lazyloader.zep:
```diff
@@ -28,16 +30,21 @@ class LazyLoader
 	/**
 	 * Phalcon\Mvc\Micro\LazyLoader constructor
+	 *
+	 * @param \Phalcon\Mvc\Model\BinderInterface modelBinder
```
not needed

sergeyklay commented on phalcon/mvc/micro/lazyloader.zep:
```diff
@@ -59,6 +66,13 @@ class LazyLoader
+		let modelBinder = this->_modelBinder;
+
+		if modelBinder != null {
+			let bindCacheKey = "binding_" . definition . "_" . method;
```
the same

sergeyklay commented on phalcon/mvc/model/binder.zep:
```diff
@@ -0,0 +1,197 @@
+	/**
+	 * Phalcon\Mvc\Model\Binder constructor
+	 *
+	 * @param \Phalcon\Cache\BackendInterface cache
```
not needed

sergeyklay commented on phalcon/mvc/model/binder.zep:
```diff
@@ -0,0 +1,197 @@
+
+			return this->getParamsFromReflection(handler, params, cacheKey, methodName);
+		}
+		else {
```
else not needed. just throw Exception if handler is not Closure and methodName is null

sergeyklay commented on phalcon/mvc/model/binder.zep:
```diff
@@ -0,0 +1,197 @@
+
+	/**
+	 * Sets cache instance
+	 * @param \Phalcon\Cache\BackendInterface cache
```
not needed

sergeyklay commented on phalcon/mvc/model/binder.zep:
```diff
@@ -0,0 +1,197 @@
+	 * Sets cache instance
+	 * @param \Phalcon\Cache\BackendInterface cache
+	 */
+	public function setCache(<BackendInterface> cache)
```
```
-> ...
```

sergeyklay commented on phalcon/mvc/model/binder.zep:
```diff
@@ -0,0 +1,197 @@
+
+	/**
+	 * Gets cache instance
+	 * @return Phalcon\Cache\BackendInterface
```
not needed

Jurigag replied:
It's needed, otherwise in generated stubs it would return `Phalcon\Mvc\Model\BackendInterface`

sergeyklay replied:
just use `-> ...`

sergeyklay commented on phalcon/mvc/model/binder.zep:
```diff
@@ -0,0 +1,197 @@
+					}
+				}
+			}
+		}
```
cyclomatic complexity

Jurigag replied:
Don't see a way to solve it.

SidRoberts replied:
How about this?:

```php
for paramKey, methodParam in methodParams {
	let reflectionClass = methodParam->getClass();

	if !reflectionClass {
		continue;
	}

	let className = reflectionClass->getName();

	if typeof className != "string" {
		continue;
	}

	// ...
}
```

Also, what value could [`reflectionClass->getName()`](http://php.net/manual/en/reflectionclass.getname.php) be if it's not a string?

Jurigag replied:
Nice idea, thanks Sid :)

sergeyklay commented on CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+# [3.1.0](https://github.com/phalcon/cphalcon/releases/tag/v3.1.0) (2016-XX-XX)
+- Added `Phalcon\Mvc\Model\Binder`, class used for binding models to parameters in dispatcher, micro, added `Phalcon\Dispatcher::getBoundModels` and `Phalcon\Mvc\Micro::getBoundModels` to getting bound models
+- Deprecated `Phalcon\Dispatcher::setModelBinding`, use `Phalcon\Dispatcher::setModelBinder`
+- Deprecated `Phalcon\Mvc\Controller\BindModelInterface`, use `Phalcon\Mvc\Model\Binder\BindableInterface`
```
It should be in 3 PR

sergeyklay replied:
how about solit work to 3 pr?

Jurigag commented:
Any wat to have IntegrationTester and UniTester in one tester method ? :C I don't know why but it trying to execute method from IntegrationTester anyway.

sergeyklay commented on phalcon/dispatcher.zep:
```diff
@@ -322,11 +329,50 @@ abstract class Dispatcher implements DispatcherInterface, InjectionAwareInterfac
+	/**
+	 * Enable model binding during dispatch
+	 *
+	 * @param \Phalcon\Mvc\Model\BinderInterface modelBinder
```
This is not Phalcon issue.

Jurigag commented:
I broke travis or what ?

Jurigag commented:
Finally passed tests, had some weird seg fault. I needed to add some useless variable to store callMethod result and then set it to returnedValue, pretty weird @virgofx can you look here ? If it's related somehow to other try catch problems ?

sergeyklay commented on phalcon/dispatcher.zep:
```diff
@@ -321,12 +325,44 @@ abstract class Dispatcher implements DispatcherInterface, InjectionAwareInterfac
-	 * @param boolean value
 	 */
-	public function setModelBinding(boolean value)
+	public function setModelBinding(boolean value, var cache = null) -> <Dispatcher>
```
Could you please add PHPDoc with example here

sergeyklay commented on phalcon/dispatcher.zep:
```diff
@@ -321,12 +325,44 @@ abstract class Dispatcher implements DispatcherInterface, InjectionAwareInterfac
+	/**
+	 * Enable model binding during dispatch
+	 */
+	public function setModelBinder(<BinderInterface> modelBinder, var cache = null) -> <Dispatcher>
```
And here

sergeyklay commented on phalcon/dispatcher.zep:
```diff
@@ -737,6 +748,22 @@ abstract class Dispatcher implements DispatcherInterface, InjectionAwareInterfac
 	/**
+	 * Returns bound models from binder instance
+	 */
+	public function getBoundModels() -> array
```
The same

sergeyklay commented on phalcon/mvc/micro.zep:
```diff
@@ -1040,4 +1079,41 @@ class Micro extends Injectable implements \ArrayAccess
+	/**
+	 * Sets model binder
+	 */
+	public function setModelBinder(<BinderInterface> modelBinder, var cache = null) -> <Micro>
```
Example please

sergeyklay commented:
I think we need detailed description for some public methods

Jurigag commented:
Yea when there will be time i will add phpdoc examples.

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+     *
+     * @param IntegrationTester $I
+     */
+    public function testDispatcher(IntegrationTester $I)
```
could you name it in a bit more meaningful way? I'd really like to know what actual functionality is tested

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+        $dispatcher->setActionName('multiple');
+        $dispatcher->setParams(['people' => $people->cedula, 'robots' => $robot->id]);
+
+        $dispatcher->dispatch();
```
this is the second time you perform action. It is a good idea to make test to have 3 steps: arrange, act, assert. You already did assertion in this test, so this can be made as another test.

This is important because smaller tests are easier to read, understand, and fix. If setup part of those tests are similar you can move it to separate method and call it using `@before` annotation

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+        $dispatcher->setActionName('view');
+        $dispatcher->setParams(['people' => $people->cedula]);
+
+        $dispatcher->dispatch();
```
to separate test

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+        $dispatcher->setActionName('multiple');
+        $dispatcher->setParams(['people' => $people->cedula, 'robots' => $robot->id]);
+
+        $dispatcher->dispatch();
```
to separate test

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+        $dispatcher->setParams(['people' => $people->cedula]);
+
+        try {
+            $dispatcher->dispatch();
```
to separate test... Probably it should named like "testDispatcherInstanceException" also you can use something 

```
$I->expectException()
```
[method](http://codeception.com/docs/modules/Asserts#expectException) in this test

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+        }
+
+        $dispatcher->setModelBinder($this->modelBinder);
+        $dispatcher->dispatch();
```
should be another test

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+    }
+
+    /**
+     * Tests micro with handlers and model binder
```
don't use comments to describe the goal of a test. Use a test name for that:

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+     *
+     * @param IntegrationTester $I
+     */
+    public function testMicroHandler(IntegrationTester $I)
```
`public function testMicroHandlerAndModelBinder`

anyway I think you can make the test name even more verbose and specific

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+        $I->assertEquals([
+            'people' => 'Phalcon\\Test\\Models\\People',
+        ], $this->cache->get('_PHMB_/{people}'));
+    }
```
yeah, this test can be split into few as well

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+     *
+     * @param IntegrationTester $I
+     */
+    public function testMicroController(IntegrationTester $I)
```
testMicroControllerAndModelBinder

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+        $I->assertEquals([
+            'people' => 'Phalcon\\Test\\Models\\People',
+        ], $this->cache->get('_PHMB_Test9Controller_viewAction'));
+    }
```
should be split into few tests... each action section should make its test

DavertMik commented on tests/integration/Mvc/Model/BinderCest.php:
```diff
@@ -0,0 +1,500 @@
+        $I->assertEquals([
+            'people' => 'Phalcon\\Test\\Models\\People'
+        ], $this->cache->get('_PHMB_Test9Controller_viewAction'));
+    }
```
just the same: split it!

Jurigag commented:
Yea on weekend i will try to refactor tests. Most of work is done but don't have time to finish it :)

Jurigag commented:
@DavertMik @sergeyklay something like this tests now ?

sergeyklay commented on phalcon/mvc/micro/lazyloader.zep:
```diff
@@ -43,25 +47,48 @@ class LazyLoader
 	/**
 	 * Initializes the internal handler, calling functions on it
 	 *
+	 * @todo move code to callMethod method in 4.0.0 and remove this method
```
Refs: https://github.com/phalcon/cphalcon/issues/12098

sergeyklay commented on phalcon/mvc/model/binder/bindableinterface.zep:
```diff
@@ -0,0 +1,34 @@
+ +------------------------------------------------------------------------+
+ | Phalcon Framework                                                      |
+ +------------------------------------------------------------------------+
+ | Copyright (c) 2011-2016 Phalcon Team (https://phalconphp.com)       |
```
Could you please fix indents?

sergeyklay commented on phalcon/mvc/model/binder/bindableinterface.zep:
```diff
@@ -0,0 +1,34 @@
+	/**
+	 * Return the model name or models names and parameters keys associated with this class
+	 */
+	abstract function getModelName() -> string|array;
```
`string | array`

sergeyklay commented on phalcon/mvc/model/binder.zep:
```diff
@@ -0,0 +1,196 @@
+	/**
+	 * Gets cache instance
+	 */
+	public function setCache(<BackendInterface> cache) -> <Binder>
```
`-> <BinderInterface>`

Jurigag commented:
I removed this @todo comment and added comment in proper issue. I think it's ready to merge @sergeyklay

Those tests are fine now ? You meant something like this ? @DavertMik

sergeyklay commented:
Thanks
